### PR TITLE
Fixes issue/3571-auto-login-sitepicker-changes

### LINKED ISSUE #3571: [UL&S] Remove "Pick store to connect" from store address login flow
When someone logs in with the “enter your store address” flow, they should be taken directly to the My Store section and should not see the "Pick a store to connect" screen. Can we remove it?

Screenshot of the "Select Store" screen showing the "Pick store to connect" option during login:

|IMG_3937|
|---|
|<sup>Tested with WCiOS 5.9.0.0 TestFlight beta, iPhone 6S, iOS 13.7. Note: ignore the user icon, it's a PNG with a black image on a transparent background. :)</sup>|

<sup>(internal reference: p5T066-1K4-p2#comment-6955 ht @Garance91540)</sup>

### PR DESCRIPTION
**Pull request by anitaa1990**
Closes: #3571 

## Current Implementation
When someone logs in with the “enter your store address” flow, they should be taken directly to the My Store section and should not see the "Pick a store to connect" screen again. 

## Changes
- There are some changes made to the login lib. This PR updates the WordPresAuthenticator lib.
- Adds two new methods to `AuthenticationManager`:
   - `shouldStoreLoginSiteAddress()` – This always returns true if we want to store the site address in `UserDefaults`, after the site address verification is successful.
   - `storeLoginSiteAddress()` – This method captures the site address and can be used by the Woo app to store this address in `UserDefaults`.
- Adds a new key to `UserDefaults` -> `loginSiteAddress`.
- Adds logic in the `SitePickerViewController` to fetch this stored site address from `UserDefaults` after we fetch the list of sites for the account. Then matches the site corresponding to this site address. If it's a match, automatically initiate the site picker to login to that site, without having the user need to click on the `Continue` button.

### Notes
- The logic added in this PR is pretty much the same as Android. While we might want to, at a later stage, update this logic so we don't need to display the Site picker at all, this is currently not possible in Android without making extensive changes. So I followed the same logic in iOS too :) 
- This is the first time I've opened a PR in iOS so I'm sure there will be tons of corrections so feel free to add any suggestions or improvements to the code :) 
- I have not added any unit tests for this. I will look into the existing unit tests and update them if it is applicable.

##### Login with Site address
- Logout of the app.
- Click on the `Continue with Site address` button.
- Enter a valid site address, email + password.
- Notice that in the `Site picker` screen you are automatically redirected to the store that you entered in step one. 

##### Login with Site address - failure case
- Logout of the app.
- Click on the `Continue with Site address` button.
- Enter a valid site address but invalid email or password.
- Notice that nothing changes from previous error handling.

##### Login with WordPress.com
- Logout of the app.
- Click on the `Continue with WordPress.com` button.
- Enter a valid email + password.
- Notice that in the `Site picker` screen you are asked to select a store before being  redirected to the store that you entered.

##### Other scenarios
- Try logging in with different options from the app. Namely, 
	- `Continue with Apple`
	- `Continue with google`
	
---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by jaclync** [comment]
@anitaa1990 nice job making this improvement and figuring out WPAuthenticator development!

Sorry I hadn't looked into the issue too much before you made the authenticator changes, which probably took some time to set up 😓 I think there is another way to determine if the user logged in with a site address without making changes to the WPAuthenticator.

In `AuthenticationManager.presentLoginEpilogue` callback after the auth flow and before we present the store picker, we can actually use `AuthenticatorCredentials.wpcom?.siteURL` to get the site URL. If the user logs in with a site address, this is set to the input site address. Otherwise, if the user logs in with WordPress.com credentials without a site address, this URL is set to a placeholder value `https://wordpress.com` (I'm not sure why this isn't just `nil`) in the WPAuthenticator.

In [this WIP branch](https://github.com/woocommerce/woocommerce-ios/compare/issue/3571-autologin-site-changes-without-authenticator?expand=1), I made some changes to test out this check and it seemed to work. Lemme know what you think in case I missed something! The remaining work on the WIP branch is to add unit tests on `ULAccountMatcher`

**Comment by anitaa1990** [comment]
Thanks for the quick review @jaclync 🙇‍♀️ ! 

> In AuthenticationManager.presentLoginEpilogue callback after the auth flow and before we present the store picker, we can actually use AuthenticatorCredentials.wpcom?.siteURL to get the site URL. If the user logs in with a site address, this is set to the input site address. Otherwise, if the user logs in with WordPress.com credentials without a site address, this URL is set to a placeholder value https://wordpress.com (I'm not sure why this isn't just nil) in the WPAuthenticator.

I did look at this code first before setting up the Authenticator changes 😅 But I noticed, same as you, that the placeholder value for the URL was `https://wordpress.com` and I couldn't think of a way to differentiate between site url login and other flows by parsing the `https://wordpress.com` value. What if this value changed in the future and we were using it to differentiate this flow? (That could be overthinking on my part 😅 ).

But I didn't look into why `wordpress.com` placeholder was returned and didn't think about `ULAccountMatcher` 😓 If it's possible to return `NIL` here, then this is a much cleaner solution indeed 👍 I'll look at it a bit more and update this PR 😊 Thank you once again for the WIP work and the setup you had to do beforehand!

**Comment by jaclync** [comment]
> I did look at this code first before setting up the Authenticator changes 😅 But I noticed, same as you, that the placeholder value for the URL was `https://wordpress.com` and I couldn't think of a way to differentiate between site url login and other flows by parsing the `https://wordpress.com` value. What if this value changed in the future and we were using it to differentiate this flow? (That could be overthinking on my part 😅 ).
> 


> But I didn't look into why `wordpress.com` placeholder was returned and didn't think about `ULAccountMatcher` 😓 If it's possible to return `NIL` here, then this is a much cleaner solution indeed 👍 I'll look at it a bit more and update this PR 😊 Thank you once again for the WIP work and the setup you had to do beforehand!



After testing login with WP.com account with a breakpoint, it looks like `https://wordpress.com` site URL is set [here](https://github.com/wordpress-mobile/WordPressAuthenticator-iOS/blob/a8ba0d9bd835ba8dad2c99c734905341092e5fb7/WordPressAuthenticator/Credentials/WordPressComCredentials.swift#L30-L34). We could update it to set to an empty value or `nil` in the authenticator, but that might affect WPiOS. Since iOS only supports WP.com login with/without a site address, I think it's safe to assume `https://wordpress.com` is from login without a site address. Alternatively, we could also add a new field to `WordPressComCredentials` for **user input** site address and do the check on `https://wordpress.com` there but it might be some extra work 😅

**Comment by anitaa1990** [comment]
Hey @jaclync, Based on your recommendation, I updated this PR to update `ULAccountMatcher`, which seemed like the most logical solution (compared to updating the WordPressAuthenticator 😅 ).

File: Podfile
Comment (jaclync) [inline Podfile]: nit: it'd be nice not having diffs on the `Podfile` when there are no pod changes, since any diffs change the checksum in `Podfile.lock` below 🙂 If you revert the change on this line then run `bundle exec pod install` again, the diff `Podfile.lock` should be back to the value in `trunk` too

```suggestion
  # pod 'WordPressAuthenticator', :path => '../WordPressAuthenticator-iOS'
```

**Review by jaclync — APPROVED** [review]
Thanks for fixing this UX @anitaa1990! 💟 Code and main testing steps ✅ 

For the additional Google and Apple login, I did notice that the site picker still showed afterward. I debugged a bit, and it requires some changes in WPAuthenticator which takes more testing with and without 2FA. Since most people still sign in with WP.com account after checking the event `woocommerceios_unified_login_step` in Tracks, I think we can merge this PR first and create a separate issue or leave another TODO/subtask in the original issue to fix third-party login. Lemme know what you think!

**Comment by anitaa1990** [comment]
That's a great catch @jaclync! I'll open a separate issue for that and take a look in the next week :) Unrelated ❓, it looks like the build tests are failing but it is working locally. Is there something I need to do or something I did wrong here 😓 ?  

I feel like it might have something to do the `Podfile.lock`. Should I revert that change and try again?

**Comment by jaclync** [comment]
> I'll open a separate issue for that and take a look in the next week :)

Sounds good thank you! I think we can try fixing it by setting `GoogleAuthViewController` (and the Apple equivalent view controller)'s `loginFields` from its caller view controller. From my testing, it was `GetStartedViewController` calling `GoogleAuthViewController.instantiate`. There might be other entry points to the Google auth VC, we can do a search on all the `GoogleAuthViewController.instantiate` use cases.

> Unrelated ❓, it looks like the build tests are failing but it is working locally. Is there something I need to do or something I did wrong here 😓 ? I feel like it might have something to do the Podfile.lock. Should I revert that change and try again?

Yea it looks like the CI failed to build from Cocoapods setup, and reverting the change should fix this build issue. The PR still has a diff on the checksum in `Podilfe.lock`, which doesn't match the `Podfile` from `trunk` anymore. If you run `bundle exec pod install` locally, that should also generate a diff in `Podilfe.lock` to revert the change 😄

**Comment by anitaa1990** [comment]
> If you run bundle exec pod install locally, that should also generate a diff in Podilfe.lock to revert the change 😄

Oh duh! I did that and it worked but I never committed that change 🤦‍♀️ ! I'll do that now and then wait for CI to 🟢 before merging. Thanks @jaclync!!

**Comment by anitaa1990** [comment]
> I think we can try fixing it by setting GoogleAuthViewController (and the Apple equivalent view controller)'s loginFields from its caller view controller. From my testing, it was GetStartedViewController calling GoogleAuthViewController.instantiate. There might be other entry points to the Google auth VC, we can do a search on all the GoogleAuthViewController.instantiate use cases.

Thanks for the pointers 😄

**Pull request merged**
